fdbclient/PaxosConfigTransaction.actor.cpp

```diff
@@ -18,118 +18,269 @@
+		wait(waitForAll(getGenerationFutures));
+		return getGenerationFutures[0].get().generation;
+	}
+
+	ACTOR static Future<Optional<Value>> get(PaxosConfigTransactionImpl* self, Key key) {
+		if (!self->getGenerationFuture.isValid()) {
+			self->getGenerationFuture = getGeneration(self);
+		}
+		state ConfigKey configKey = ConfigKey::decodeKey(key);
+		ConfigGeneration generation = wait(self->getGenerationFuture);
+		// TODO: Load balance
+		ConfigTransactionGetReply reply =
+		    wait(self->ctis[0].get.getReply(ConfigTransactionGetRequest{ generation, configKey }));
+		if (reply.value.present()) {
+			return reply.value.get().toValue();
+		} else {
+			return Optional<Value>{};
+		}
+	}
+
+	ACTOR static Future<RangeResult> getConfigClasses(PaxosConfigTransactionImpl* self) {
+		if (!self->getGenerationFuture.isValid()) {
+			self->getGenerationFuture = getGeneration(self);
+		}
+		ConfigGeneration generation = wait(self->getGenerationFuture);
+		// TODO: Load balance
+		ConfigTransactionGetConfigClassesReply reply =
+		    wait(self->ctis[0].getClasses.getReply(ConfigTransactionGetConfigClassesRequest{ generation }));
+		RangeResult result;
+		result.reserve(result.arena(), reply.configClasses.size());
+		for (const auto& configClass : reply.configClasses) {
+			result.push_back_deep(result.arena(), KeyValueRef(configClass, ""_sr));
+		}
+		return result;
+	}
+
+	ACTOR static Future<RangeResult> getKnobs(PaxosConfigTransactionImpl* self, Optional<Key> configClass) {
+		if (!self->getGenerationFuture.isValid()) {
+			self->getGenerationFuture = getGeneration(self);
+		}
+		ConfigGeneration generation = wait(self->getGenerationFuture);
+		// TODO: Load balance
+		ConfigTransactionGetKnobsReply reply =
+		    wait(self->ctis[0].getKnobs.getReply(ConfigTransactionGetKnobsRequest{ generation, configClass }));
+		RangeResult result;
+		result.reserve(result.arena(), reply.knobNames.size());
+		for (const auto& knobName : reply.knobNames) {
+			result.push_back_deep(result.arena(), KeyValueRef(knobName, ""_sr));
+		}
+		return result;
+	}
+
+	ACTOR static Future<Void> commit(PaxosConfigTransactionImpl* self) {
+		if (!self->getGenerationFuture.isValid()) {
+			self->getGenerationFuture = getGeneration(self);
+		}
+		wait(store(self->toCommit.generation, self->getGenerationFuture));
+		self->toCommit.annotation.timestamp = now();
+		std::vector<Future<Void>> commitFutures;
+		commitFutures.reserve(self->ctis.size());
+		for (const auto& cti : self->ctis) {
+			commitFutures.push_back(cti.commit.getReply(self->toCommit));
+		}
+		// FIXME: Must tolerate failures and disagreement
+		wait(quorum(commitFutures, commitFutures.size() / 2 + 1));
+		self->committed = true;
+		return Void();
+	}
+
+public:
+	Future<Version> getReadVersion() {
+		if (!getGenerationFuture.isValid()) {
+			getGenerationFuture = getGeneration(this);
+		}
+		return map(getGenerationFuture, [](auto const& gen) { return gen.committedVersion; });
```

Comment:
Should this be `liveVersion`? I may be getting a little confused, but my understanding is `committedVersion` is the latest version that has been committed on the coordinator, while `liveVersion` is incremented each time a new generation request is made to the coordinator and is what is used to determine whether a transaction can commit or if a conflicting transaction has already started. At the very least, shouldn't the versions in `getReadVersion` and `getCachedReadVersion` be the same?

Reply:
You're right, this should return `get.liveVersion`.